File: Final_Project/Code/Project_4/Bai_toan_5/Exercises/3/CSS_bt5_ex3.cpp
#include <iostream>
#include <vector>
using namespace std;

class Graph {
public:
    int n;                       // so dinh
    vector<vector<int>> adj;     // danh sach ke

    Graph(int vertices) : n(vertices) {
        adj.resize(n);
    }

    void add_edge(int u, int v) {
        // Them canh (u,v)
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    void generate_path() {
        // Sinh do thi duong Pn
        for (int i = 0; i < n - 1; ++i) {
            add_edge(i, i + 1);
        }
    }

    void generate_circle() {
        // Sinh do thi vong Cn
        generate_path();
        if (n > 2) {
            add_edge(0, n - 1);
        }
    }

    void generate_wheel() {
        // Sinh do thi banh xe Wn
        generate_circle();
        for (int i = 0; i < n - 1; ++i) {
            add_edge(n - 1, i);
        }
    }
};

int main() {
    Graph g(5);
    g.generate_path();
    cout << "Path graph edges: ";
    for (int u = 0; u < g.n; ++u) {
        for (int v : g.adj[u]) {
            if (u < v) cout << "(" << u << "," << v << ") ";
        }
    }
    cout << endl;
    return 0;
}

File: Final_Project/Code/Project_4/Bai_toan_5/Exercises/3/CSS_bt5_ex3.py
class Graph:
    def __init__(self, n):
        # n: so dinh
        # adj: danh sach ke
        self.n = n
        self.adj = [[] for _ in range(n)]

    def add_edge(self, u, v):
        # Them canh (u,v)
        self.adj[u].append(v)
        self.adj[v].append(u)

    def generate_path(self):
        # Sinh do thi duong Pn
        for i in range(self.n - 1):
            self.add_edge(i, i + 1)

    def generate_circle(self):
        # Sinh do thi vong Cn
        self.generate_path()
        if self.n > 2:
            self.add_edge(0, self.n - 1)

    def generate_wheel(self):
        # Sinh do thi banh xe Wn
        self.generate_circle()
        center = self.n - 1
        for i in range(self.n - 1):
            self.add_edge(center, i)

# Vi du su dung
if __name__ == "__main__":
    g = Graph(5)
    g.generate_path()
    print("Path graph edges:", [(u, v) for u in range(g.n)
                                for v in g.adj[u] if u < v])
    g = Graph(5)
    g.generate_circle()
    print("Circle graph edges:", [(u, v) for u in range(g.n)
                                  for v in g.adj[u] if u < v])
    g = Graph(5)
    g.generate_wheel()
    print("Wheel graph edges:", [(u, v) for u in range(g.n)
                                 for v in g.adj[u] if u < v])